proxy/src/auth/backend/jwt.rs

```diff
@@ -0,0 +1,549 @@
+use serde::de::DeserializeOwned;
+use signature::Verifier;
+use tokio::time::Instant;
+
+use crate::intern::EndpointIdInt;
+
+#[derive(Default)]
+pub struct JWKCache {
+    client: reqwest::Client,
+
+    map: DashMap<EndpointIdInt, Arc<JWKCacheEntryLock>>,
+}
+
+pub struct JWKCacheEntryLock {
+    cached: ArcSwapOption<JWKCacheEntry>,
+    lookup: tokio::sync::Semaphore,
+}
+
+impl Default for JWKCacheEntryLock {
+    fn default() -> Self {
+        JWKCacheEntryLock {
+            cached: ArcSwapOption::empty(),
+            lookup: tokio::sync::Semaphore::new(1),
+        }
+    }
+}
+
+pub struct JWKCacheEntry {
+    /// Should refetch at least every hour to verify when old keys have been removed.
+    /// Should refetch when new key IDs are seen only every 5 minutes or so
+    last_retrieved: Instant,
+
+    // /// jwks urls
+    // urls: Vec<url::Url>,
+    /// cplane will return multiple JWKs urls that we need to scrape.
+    key_sets: ahash::HashMap<url::Url, jose_jwk::JwkSet>,
+}
+
+/// How to get the JWT auth rules
+pub trait FetchAuthRules: Clone + Send + Sync + 'static {
+    fn fetch_auth_rules(&self) -> impl Future<Output = anyhow::Result<AuthRules>> + Send;
+}
+
+#[derive(Clone)]
+struct FetchAuthFromCplane {
+    #[allow(dead_code)]
+    endpoint: EndpointIdInt,
+}
+
+impl FetchAuthRules for FetchAuthFromCplane {
+    async fn fetch_auth_rules(&self) -> anyhow::Result<AuthRules> {
+        Err(anyhow::anyhow!("not yet implemented"))
+    }
+}
+
+pub struct AuthRules {
+    jwks_urls: Vec<url::Url>,
+}
+
+impl JWKCacheEntryLock {
+    async fn acquire_permit(&self) -> DetachedPermit {
+        match self.lookup.acquire().await {
+            Ok(permit) => {
+                permit.forget();
+                DetachedPermit
+            }
+            Err(_) => panic!("semaphore should not be closed"),
+        }
+    }
+
+    fn try_acquire_permit(&self) -> Option<DetachedPermit> {
+        match self.lookup.try_acquire() {
+            Ok(permit) => {
+                permit.forget();
+                Some(DetachedPermit)
+            }
+            Err(tokio::sync::TryAcquireError::NoPermits) => None,
+            Err(tokio::sync::TryAcquireError::Closed) => panic!("semaphore should not be closed"),
+        }
+    }
+
+    async fn renew_jwks<F: FetchAuthRules>(
+        &self,
+        permit: DetachedPermit,
+        client: &reqwest::Client,
+        auth_rules: &F,
+    ) -> anyhow::Result<Arc<JWKCacheEntry>> {
+        let _permit = permit.attach(&self.lookup);
+
+        // double check that no one beat us to updating the cache.
+        let now = Instant::now();
+        let guard = self.cached.load_full();
+        if let Some(cached) = guard {
+            let last_update = now.duration_since(cached.last_retrieved);
+            if last_update < Duration::from_secs(300) {
+                return Ok(cached);
+            }
+        }
+
+        let rules = auth_rules.fetch_auth_rules().await?;
```

Comment:
Where are these auth rules defined? Should the code be more resilient against malformed rules? Maybe leave a TODO comment.

Reply:
For auth broker, they are returned by cplane. For local proxy, they are configured by neonctl/config. The rules we are intending to support are just
1. signature is correct (using the JWKs urls for the signature keys)
2. audience is correct

Eventually we might support more. At this stage there's not much to validate besides checking that it parses correctly.

Reply:
Hmm. That's fair. We will plan to run some validations of the URL in console as well FYI to make sure it doesn't get saved if it's not an expected JWKs URL.

@davidgomes does it sound good to you that we should try and remove query params and ports from the URL string as well? Strictly just https+host+path

Reply:
Yes! I've added that to https://github.com/neondatabase/cloud/issues/16284#issuecomment-2286426288.

Reply:
Ok, sounds good. Maybe add a TODO referencing cloud#16284.